diff --git a/vendor/magento/module-company/Plugin/Sales/Model/ResourceModel/Order/Grid/CollectionPlugin.php b/vendor/magento/module-company/Plugin/Sales/Model/ResourceModel/Order/Grid/CollectionPlugin.php
index de7abbaf5..3988dbbcb 100644
--- a/vendor/magento/module-company/Plugin/Sales/Model/ResourceModel/Order/Grid/CollectionPlugin.php
+++ b/vendor/magento/module-company/Plugin/Sales/Model/ResourceModel/Order/Grid/CollectionPlugin.php
@@ -23,14 +23,16 @@ class CollectionPlugin
         $printQuery = false,
         $logQuery = false
     ) {
-        $subject->getSelect()
-            ->joinLeft(
-                ['company_order' => $subject->getTable(
-                    \Magento\Company\Setup\InstallSchema::ORDER_ENTITY_TABLE_NAME
-                )],
-                'main_table.entity_id = company_order.order_id',
-                ['company_name']
-            );
+        if (!$subject->isLoaded()) {
+            $subject->getSelect()
+                ->joinLeft(
+                    ['company_order' => $subject->getTable(
+                        \Magento\Company\Setup\InstallSchema::ORDER_ENTITY_TABLE_NAME
+                    )],
+                    'main_table.entity_id = company_order.order_id',
+                    ['company_name']
+                );
+        }
 
         return [$printQuery, $logQuery];
     }
