diff --git a/vendor/magento/module-google-analytics/view/frontend/web/js/google-analytics.js b/vendor/magento/module-google-analytics/view/frontend/web/js/google-analytics.js
index cd6292b39e9..60a131ae746 100644
--- a/vendor/magento/module-google-analytics/view/frontend/web/js/google-analytics.js
+++ b/vendor/magento/module-google-analytics/view/frontend/web/js/google-analytics.js
@@ -53,7 +53,7 @@ define([
             }
 
             // Process orders data
-            if (config.ordersTrackingData.length) {
+            if (config.ordersTrackingData.hasOwnProperty('currency')) {
                 ga('require', 'ec', 'ec.js');
 
                 //Set currency code
