Index: vendor/magento/module-ui/Component/MassAction/Filter.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- a/vendor/magento/module-ui/Component/MassAction/Filter.php	(revision Local Version)
+++ b/vendor/magento/module-ui/Component/MassAction/Filter.php	(revision Shelved Version)
@@ -105,6 +105,8 @@
                 $collection->getIdFieldName(),
                 ['in' => $idsArray]
             );
+        } else {
+            throw new LocalizedException(__('Please clean filters before applying MassActions.'));
         }
         return $collection;
     }
Index: vendor/magento/module-ui/view/base/web/js/grid/columns/multiselect.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- a/vendor/magento/module-ui/view/base/web/js/grid/columns/multiselect.js	(revision Local Version)
+++ b/vendor/magento/module-ui/view/base/web/js/grid/columns/multiselect.js	(revision Shelved Version)
@@ -52,6 +52,7 @@
 
             listens: {
                 '${ $.provider }:params.filters': 'onFilter',
+                '${ $.provider }:params.search' : 'onFilter',
                 selected: 'onSelectedChange',
                 rows: 'onRowsChange'
             },
