diff --git a/vendor/magento/module-catalog-search/Block/Advanced/Result.php b/vendor/magento/module-catalog-search/Block/Advanced/Result.php
index 76e15ecd6a5..8f75f062987 100644
--- a/vendor/magento/module-catalog-search/Block/Advanced/Result.php
+++ b/vendor/magento/module-catalog-search/Block/Advanced/Result.php
@@ -96,8 +96,12 @@ class Result extends Template
 
         $availableOrders = $category->getAvailableSortByOptions();
         unset($availableOrders['position']);
+        $availableOrders['relevance'] = __('Relevance');
 
-        $this->getChildBlock('search_result_list')->setAvailableOrders($availableOrders);
+        $this->getChildBlock('search_result_list')
+            ->setAvailableOrders($availableOrders)
+            ->setDefaultDirection('desc')
+            ->setSortBy('relevance');
     }
 
     /**
