diff --git a/vendor/magento/module-analytics/Model/Cryptographer.php b/vendor/magento/module-analytics/Model/Cryptographer.php
index 6905eee372a..4236508cf78 100644
--- a/vendor/magento/module-analytics/Model/Cryptographer.php
+++ b/vendor/magento/module-analytics/Model/Cryptographer.php
@@ -124,7 +124,11 @@ class Cryptographer
      */
     private function validateCipherMethod($cipherMethod)
     {
-        $methods = openssl_get_cipher_methods();
+         $methods = array_map(
+            'strtolower',
+            openssl_get_cipher_methods()
+        );
+        $cipherMethod = strtolower($cipherMethod);
         return (false !== array_search($cipherMethod, $methods));
     }
 }
