diff --git a/vendor/magento/module-google-tag-manager/Model/Plugin/Quote/SetGoogleAnalyticsOnCartRemove.php b/vendor/magento/module-google-tag-manager/Model/Plugin/Quote/SetGoogleAnalyticsOnCartRemove.php
index d5a5dd20620..e14fc0d3091 100644
--- a/vendor/magento/module-google-tag-manager/Model/Plugin/Quote/SetGoogleAnalyticsOnCartRemove.php
+++ b/vendor/magento/module-google-tag-manager/Model/Plugin/Quote/SetGoogleAnalyticsOnCartRemove.php
@@ -44,7 +44,9 @@ class SetGoogleAnalyticsOnCartRemove
     public function afterRemoveItem(\Magento\Quote\Model\Quote $subject, $result, $itemId)
     {
         $item = $subject->getItemById($itemId);
-        $this->setItemForTriggerRemoveEvent($this->helper, $this->registry, $item, $item->getQty());
+        if ($item) {
+            $this->setItemForTriggerRemoveEvent($this->helper, $this->registry, $item, $item->getQty());
+        }
 
         return $result;
     }
