diff --git a/vendor/magento/module-catalog-graph-ql/etc/schema.graphqls b/vendor/magento/module-catalog-graph-ql/etc/schema.graphqls
index d4b98b311fc..19ffe4be96f 100644
--- a/vendor/magento/module-catalog-graph-ql/etc/schema.graphqls
+++ b/vendor/magento/module-catalog-graph-ql/etc/schema.graphqls
@@ -9,7 +9,7 @@ type Query {
         currentPage: Int = 1 @doc(description: "Specifies which page of results to return. The default value is 1."),
         sort: ProductAttributeSortInput @doc(description: "Specifies which attributes to sort on, and whether to return the results in ascending or descending order.")
     ): Products
-    @resolver(class: "Magento\\CatalogGraphQl\\Model\\Resolver\\Products") @doc(description: "The products query searches for products that match the criteria specified in the search and filter attributes.") @cache(cacheIdentity: "Magento\\CatalogGraphQl\\Model\\Resolver\\Product\\Identity")
+    @resolver(class: "Magento\\CatalogGraphQl\\Model\\Resolver\\Products") @doc(description: "The products query searches for products that match the criteria specified in the search and filter attributes.")
     category (
      id: Int @doc(description: "Id of the category.")
     ): CategoryTree
