diff --git a/vendor/magento/framework/Archive/Tar.php b/vendor/magento/framework/Archive/Tar.php
index 7fe1255e5b8..a858b241151 100644
--- a/vendor/magento/framework/Archive/Tar.php
+++ b/vendor/magento/framework/Archive/Tar.php
@@ -4,15 +4,15 @@
  * See COPYING.txt for license details.
  */
 
+namespace Magento\Framework\Archive;
+
+use Magento\Framework\Archive\Helper\File;
+
 /**
  * Class to work with tar archives
  *
  * @author      Magento Core Team <core@magentocommerce.com>
  */
-namespace Magento\Framework\Archive;
-
-use Magento\Framework\Archive\Helper\File;
-
 class Tar extends \Magento\Framework\Archive\AbstractArchive implements \Magento\Framework\Archive\ArchiveInterface
 {
     /**
@@ -259,10 +259,7 @@ class Tar extends \Magento\Framework\Archive\AbstractArchive implements \Magento
                 );
             }
 
-            array_shift($dirFiles);
-            /* remove  './'*/
-            array_shift($dirFiles);
-            /* remove  '../'*/
+            $dirFiles = array_diff($dirFiles, ['..', '.']);
 
             foreach ($dirFiles as $item) {
                 $this->_setCurrentFile($file . $item)->_createTar();
