diff --git a/lib/web/mage/requirejs/resolver.js b/lib/web/mage/requirejs/resolver.js
index f3d4b79216a..f72d920b75e 100644
--- a/lib/web/mage/requirejs/resolver.js
+++ b/lib/web/mage/requirejs/resolver.js
@@ -28,13 +28,37 @@ define([
     }
 
     /**
+     * Checks if provided module registered during load.
+     *
+     * @param {Object} module - Module to be checked.
+     * @return {Boolean}
+     */
+    function isRegistered(module) {
+        return registry[module.id];
+    }
+
+    /**
+     * Checks if provided module is rejected during load.
+     *
+     * @param {Object} module - Module to be checked.
+     * @return {Boolean}
+     */
+    function isRejected(module) {
+        return isRegistered(module) && registry[module.id].error;
+    }
+
+    /**
      * Checks if provided module has unresolved dependencies.
      *
      * @param {Object} module - Module to be checked.
      * @returns {Boolean}
      */
     function isPending(module) {
-        return !!module.depCount;
+        if (!module.depCount) {
+            return false;
+        }
+
+        return module.depCount > _.filter(module.depMaps, isRegistered).length;
     }
 
     /**
